### ISSUE TITLE
Benchmarks showing high performance?

### ISSUE THREAD

**Issue by jzabroski** [hpeebles/CacheMeIfYouCan#1]
Curious if you have any benchmarkDotNet benchmarks or similar.

**Comment by hpeebles** [comment]
I'll have some time this weekend. I'll see what I can get sorted!

**Comment by jzabroski** [comment]
FASTER uses YCSB for benchmarking https://github.com/Microsoft/FASTER/tree/master/cc/benchmark-dir

**Comment by hpeebles** [comment]
I've added some basic benchmarks which show the overhead added on top of using System.Runtime.MemoryCache directly (which is the cache used by default) for each type of function cache (single key, enumerable key, multi param key and multi param enumerable key).

The results of these benchmarks are here -
https://github.com/hpeebles/CacheMeIfYouCan/blob/master/src/Benchmarks/Benchmarks.FunctionCacheComparison/FunctionCacheBenchmarksRunner.cs#L124-L161

From these results you can see that the 2 enumerable key function caches add a relatively large overhead in this simple example as there is a fair bit of complexity involved in their workflow. They check the cache for each of the requested keys, invoke the underlying function for any keys that are missing (if any), set the newly found values into the cache, then return the complete dictionary of values. This involves having to loop through the set of keys a few times which the baseline benchmark doesn't have to do.

Also each single key function cache allocates a fair bit of memory when testing large sets of keys. This is because each request to a function cache allocates an object on the heap to contain the results, so when executing the test for lots of keys, lots of objects get allocated on the heap in the single key function caches but not for the enumerable key function caches since they handle all the keys in a single request.

ps. A lot of what I said above may not make sense right now due to the massive lack of documentation! Documentation is the next priority. People need to know what this big lump of code is and how to use it!

**Comment by jzabroski** [comment]
Nice job.  I would put the benchmark documentation in the README.md since
BenchmarkDotNet supports generating GitHub flavored Markdown.

I understand perfectly what this does - it's half cache, half type-safe
Memoizer.  The simplest idiom would be to cache an anonymous type and use
the interface (record of properties) for that type as the key in a
key-value pair map.  I have used similar such idioms in the past.

I don't _love_ the API of your project, but it's nice to see like-minded
souls out there.

In my experience, the painful reality is that as a project reaches larger
scale, such idioms don't scale too well, principally because the dominating
factor in cache performance becomes IO throughput and that is going to be
bounded by your network's bandwidth-delay product and the size of your keys
and what you are caching.  Such fancy frameworks create really large keys,
and that's their drawback to scaling them.  On the other hand, I never seen
such frameworks provide progressive enhancements for "registering keys",
since most applications have a meta-stable list of keys (or can be
re-designed to work that way).

**Comment by hpeebles** [comment]
We use this library a lot at my work and the best solution we’ve found for really high volume scenarios is to hash the key objects to 64 bits (we have millions of keys but nowhere near enough for collisions to be an issue + this isn’t critical data) and use a Redis cluster as the main cache but then by also storing keys whose corresponding value is null in memory.

By doing it like this we get round the large key problem + our local memory usage is really efficient as we see an in memory hit ratio of around 20% while storing a few million Int64 keys and no values locally.

**Comment by jzabroski** [comment]
> We use this library a lot at my work and the best solution we’ve found for really high volume scenarios is to hash the key objects to 64 bits (we have millions of keys but nowhere near enough for collisions to be an issue

But you understand why you don't have lots of key collisions, right?  An unsigned 64 bit integer stores 0 to 18,446,744,073,709,551,615 or 18,446,744,073,709,551,616 unique values.  If you have 10 million objects in your cache, that's 10,000,000/18,446,744,073,709,551,616

```powershell
$num = 10000000
$denom = 18446744073709551616
$num / $denom
# Outputs: 0.0000000000005421010862427522
```

Even if you move to 32 bit keys, you should not have that many collisions.  Unsigned 32 bit stores 0 to 4,294,967,295 or 4,294,967,296 unique values.  That's 10,000,000/4,294,967,296

```powershell
$denom2 = 4294967296
$num / $denom2
# Outputs: 0.0023283064365387
```

YMMV, and my naive keys divided by cache slots is only a proxy to things like cache load factor and birthday paradox calculation for key collisions, but keep in mind that even a [1% cache miss ratio can cause a 10x performance penalty](http://highscalability.com/blog/2012/10/11/ramcube-exploiting-network-proximity-for-ram-based-key-value.html).

Happy to keep learning and discussing with you.

**Comment by hpeebles** [comment]
We use a 32 bit hash as the keys for one of our datasets where we know the entire dataset in advance. In that case we calculated the 32 bit hashes of each key to ensure there were no collisions.

But with our largest dataset where the data is unbounded we chose 64 bit hashes rather than 32 bit just to be safe. We have roughly 10 million keys so based on the table on this page https://en.wikipedia.org/wiki/Birthday_problem the chance of us having a collision at any given moment is less than 0.000001%. Which we are more than happy with! Also the values are far larger than the keys so the overall memory usage in Redis wouldn't be reduced all that much by using 32 bit keys rather than 64. In local memory it would be reduced since we only store keys but we settled on 64 bit in the end since the memory usage still isn't very high in local memory.

**Comment by hpeebles** [comment]
Also, out of curiosity... when you say you don't _love_ the API, would you be able to elaborate on that? Are there any bits in particular that stood out? I've been churning this out on my own so it would be really good to hear feedback from other people!

**Comment by jzabroski** [comment]
There are several axes to think about caching and memoization from.

Traditional caching API assumes unlimited space, unlimited time by default
and then evicts entries when those constraints become unwieldy. This is a
really limited way to express intent. I would say it's a lazy way, but that
would spoil my pun: The secret to expressing intent is to be lazy!

Current state: Unlimited size, unlimited time
Desired Options:
Limited size, unlimited time
Unlimited size, limited time
Don't use the cache unless real function takes >100ms to return
Mix of strategies: Look up the concept of multi-stage programming as
pioneered by Walid Taha at Rice University about 2 decades ago. Perhaps the
first 10 elements should never be cached because it's faster to calculate,
but after that we're better off building a lookup table. Can the cache
auto-tune itself?

Such desired options are non-trivial to design well: How do you handle
deserialization errors where one version of the application has reserved
one data type for a key-value pair's value, and another version has
reserved a different data type for the same key. 7 fallacies of distributed
computing, don'tchaknow. Non-trivial, but not impossible. Very doable. Yet
we still write code like we're trying to start a fire with sticks rather
than rocket feul and a spark.

The other practical aspect is how you use it. I like declaration-site
caching, since it's less typing and can leverage generic type inference to
infer and auto-generated type-safe keys:

var result = Memorizer.Memoize(new { Id = a, IncludeInactive = b}, val =>
customerRepository.Where(c => c.Id = val. Id && val.IncludeInactive? true :
c.IsActive));

This uses anonymous types, generic type inference and Func parsimoniously
to create a type-safe cache. You can also use such an interface as a facade
to cache all caching requests across your process. Build a cache of your
cache, yo. Log callermembername and provide analytics on which methods
utilize the cache. Turn off the logging and have the JIT optimize away the
logging code path for when you don't need it. Look at how the Jikes
Recursive Virtual Machine works as inspiration for thinking about how to
have your code do your thinking for you in a meta-fantastic way.

**Comment by jzabroski** [comment]
Was that helpful?

**Comment by hpeebles** [comment]
Woah that's a lot of stuff to take in! I'm off on holiday tomorrow and am not ready for it at all (woops haha!) so haven't had a proper chance to look over this yet. But some of the stuff you mention seems really cool, especially the self tuning cache.

I started building this library as a solution to a common problem we had at work: we send a lot of requests to external sources (databases, web services, etc) and wanted a simple way to add a caching layer to improve performance and reduce load on the external resource. In these scenarios the data will be slowly changing over time so one thing I think would be cool to add is a way to automatically adjust the cache times to live based on a configured tolerance for change. As in, with internal data that we own we push out notifications when it changes so that we can remove the now out of date cache keys but for external services we cannot do this so the only way to tell if the data has changed or not is to request it again and compare it to what we have cached. Then the system would automatically adjust the times to live to try to ensure that only 1% (or whatever limit you choose) of the cached data is ever out of date.

Regarding your comment about keys being used for 2 different types, the solution offered in this library is to add a custom keyspace prefix to each function's set of keys when using a distributed cache (there's a method called WithKeyspacePrefix). This is not an issue when using a local cache since each function gets its own cache (this may change at a later date so that functions could share cached data if the user wants that).

Anyway I'm back home on the 1st. I'll take a better look at your comment then. Thanks a lot for all the comments and advice!
